src/value.cpp

```diff
@@ -516,12 +518,28 @@ MemRef::getDimsAndLayoutAndElemTy(
 
 pair<expr, expr> MemRef::load(const vector<expr> &indices) {
   expr idx = to1DIdxWithLayout(indices);
-  return m->load(bid, offset + idx);
+  auto [expr, success] = m->load(bid, offset + idx);
+
+  // check whether indices are inbound.
+  for (int i = 0; i < indices.size(); i ++)
+    success = success && z3::ult(indices[i], getDim(i));
+
```

Comment:
This may contradicts with #32 (comment).
But in mlir documents (https://mlir.llvm.org/docs/Dialects/Builtin/#index-map), the index space is always non-negative and integral. So it is not allowed to access with negative Index.

And dimension-wise index bounds are strictly checked in MLIR.
mlir::boundCheckLoadOrStoreOp in mlir/lib/Analysis/Utils.cpp may relate with this.
So we need this bound checks in well-defineness.

Reply:
Can we split this change as a separate PR?